src/models/journals.ts

```diff
@@ -0,0 +1,189 @@
+import {
+  connection,
+  Schema,
+  Document,
+  Model,
+  model,
+  Types,
+  PreSaveMiddlewareFunction,
+} from "mongoose";
+import {
+  isValidTransactionKey,
+  ITransaction,
+  transactionModel,
+} from "./transactions";
+import type { Book } from "../Book";
+import { handleVoidMemo } from "../helper/handleVoidMemo";
+import type { IAnyObject } from "../IAnyObject";
+import type { IOptions } from "../IOptions";
+
+export interface IJournal {
+  _id: Types.ObjectId;
+  datetime: Date;
+  memo: string;
+  _transactions: Types.ObjectId[] | ITransaction[];
+  book: string;
+  voided: boolean;
+  void_reason: string;
+  approved: boolean;
+}
+
+const journalSchema = new Schema<IJournal>(
+  {
+    datetime: Date,
+    memo: {
+      type: String,
+      default: "",
+    },
+    _transactions: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Medici_Transaction",
+      },
+    ],
+    book: String,
+    voided: {
+      type: Boolean,
+      default: false,
+    },
+    void_reason: String,
+    approved: {
+      type: Boolean,
+      default: true,
+    },
+  },
+  { id: false, versionKey: false, timestamps: false }
+);
+
+function processMetaField(key: string, val: unknown, meta: IAnyObject): void {
+  isValidTransactionKey(key) ? undefined : (meta[key] = val);
+}
+
+const voidJournal = async function (
+  book: Book,
+  reason: undefined | null | string,
+  options: IOptions
+) {
+  if (this.voided === true) {
+    throw new Error("Journal already voided");
+  }
+
+  reason = handleVoidMemo(reason, this.memo);
+
+  // Set this to void with reason and also set all associated transactions
+  this.voided = true;
+  this.void_reason = reason;
+
+  await this.save(options);
+
+  const transactions = await transactionModel
+    .find(
+      {
+        _journal: this._id,
+      },
+      undefined,
+      options
+    )
+    .exec();
+
+  for (let i = 0, il = transactions.length; i < il; i++) {
+    transactions[i].voided = true;
+    transactions[i].void_reason = this.void_reason;
+  }
+
+  await Promise.all(
+    transactions.map((tx) => new transactionModel(tx).save(options))
+  );
+
+  const entry = book.entry(reason, null, this._id);
+
+  for (const trans of transactions) {
+    const meta: IAnyObject = {};
+    Object.keys(trans.toObject()).forEach((key) => {
```

Comment:
I think the whole conversion makes no sense in 99.9999999999999% of all cases.

The only case were I could think of a potential use case, was when you for some reason migrate entries on the fly. But why would you do this anyway?

This could be a simple const meta = trans.meta;

I did not remove this code, as the performance is not touched by these lines as the slowest part is actually the mongoDB and not this code.

But theoretically removing this, would simplify the code.

Reply:
Meh, not a performance issue imho. Better not touch this for now.